modules/apps/frontend-js/frontend-js-a11y-web/src/main/resources/META-INF/resources/useFilteredViolations.tsx

```diff
@@ -0,0 +1,181 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import React, {Dispatch, createContext, useContext, useReducer} from 'react';
+
+import type {ImpactValue, Result} from 'axe-core';
+
+export const TYPES = {
+	CATEGORY_ADD: 'CATEGORY_ADD',
+	CATEGORY_REMOVE: 'CATEGORY_REMOVE',
+	IMPACT_ADD: 'IMPACT_ADD',
+	IMPACT_REMOVE: 'IMPACT_REMOVE',
+} as const;
+
+type TAction = {
+	payload: {value: string};
+	type: 'CATEGORY_ADD' | 'CATEGORY_REMOVE' | 'IMPACT_ADD' | 'IMPACT_REMOVE';
+};
+
+type TState = {
+	filteredViolations: Array<Result>;
+	selectedCategories: Array<String> | [];
+	selectedImpact: Array<ImpactValue> | [];
+	violations: Array<Result>;
+};
+
+const INITIAL_STATE: TState = {
+	filteredViolations: [],
+	selectedCategories: [],
+	selectedImpact: [],
+	violations: [],
+};
+
+function filterByCategories(
+	receivedTags: Array<String>,
+	selectedCategories: Array<String>
+) {
+	return selectedCategories.some((category) =>
+		receivedTags.includes(category)
+	);
+}
+
+function filterByImpact(
+	receivedImpact: ImpactValue,
+	selectedImpact: Array<ImpactValue>
+) {
+	return selectedImpact.includes(receivedImpact);
+}
+
+function addItem(list: Array<any>, value: any) {
+	return [...list, value];
+}
+
+function removeItem(list: Array<any>, value: any) {
+	return list.filter((currentItem) => currentItem !== value);
+}
+
+function violationsReducer(state = INITIAL_STATE, action: TAction) {
+	const {selectedCategories, selectedImpact, violations} = state;
+
+	const {value} = action.payload;
+
+	switch (action.type) {
+		case TYPES.CATEGORY_ADD: {
+			const newSelectedCategories = addItem(selectedCategories, value);
+
+			return {
+				...state,
+				filteredViolations: newSelectedCategories.length
+					? violations.filter(({tags}) => {
+							if (tags) {
+								return filterByCategories(
+									tags,
+									newSelectedCategories
+								);
+							}
+					  })
+					: violations,
```

Comment:
Hey @diegonvs well I think we can have a lot of problems with filters here, ideally making the filter and saving it in the state can always generate problems of state synchronization, in this case the best would be to make the filter at render time which will be safer, same filter logic within a `useMemo` that is triggered by changes to `violations`, `selectedCategories` or `selectedImpact`.

I can see some problems with this approach:

- When `violations` is updated we will not filter the data
- When adding a category filter we will make the category only filter instead of preserving the `impact` filter, the same would be the other way around

Reply:
Done! Move filteredViolations from SSOT to be filtered using `useMemo` on `FilteredViolationsContextProvider` and maintain selectedImpact/selectedCategories on SSOT